lib/Doctrine/Enumeration/Enumerated.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Enumeration;
+
+use Doctrine\Enumeration\Exception\NotSupportedException;
+use function array_search;
+use function constant;
+use function in_array;
+
+trait Enumerated
+{
+    final private function __construct()
+    {
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function validateValue($value) : void
+    {
+        if (! in_array($value, self::availableValues(), true)) {
+            throw new NotSupportedException('Invalid value.');
+        }
+    }
+
+    final public static function validateName(string $name) : void
+    {
+        if (! isset(self::availableValues()[$name])) {
+            throw new NotSupportedException('Invalid name.');
+        }
+    }
+
+    /**
+     * @param mixed $value
+     */
+    final public static function get($value) : self
+    {
+        self::validateValue($value);
+
+        $name      = array_search($value, self::availableValues(), true);
+        $instances = &self::registry(static::class);
+
+        return $instances[$name] ?? $instances[$name] = new static();
+    }
+
+    /**
+     * @return mixed
+     */
+    final private static function value(self $instance)
+    {
+        return self::availableValues()[array_search($instance, self::registry(static::class), true)];
+    }
+
+    /**
+     * @return self[]
+     */
+    private static function &registry(string $type) : array
+    {
+        static $registry = [];
+
+        if (! isset($registry[$type])) {
+            $registry[$type] = [];
+        }
+
+        $ref = &$registry[$type];
+        return $ref;
+    }
+
+    /**
+     * @return mixed[]
+     */
+    private static function availableValues() : array
+    {
+        static $values = [];
+
+        if (isset($values[static::class])) {
+            return $values[static::class];
+        }
+
+        $values[static::class] = [];
+
+        foreach ((new \ReflectionClass(static::class))->getReflectionConstants() as $name => $constantReflection) {
+            if (! $constantReflection->isPublic()) {
```

Comment:
Why only `public` constants?

Reply:
1) To allow possible private constants i.e. for maps and behavior methods, i.e.:
```php
public const FOO = 1;
public const BAR = 2;
private const DESCRIPTIONS = [
    self::FOO => 'Hello',
    self::BAR => 'World',
];

...

public function getDescription() : string
{
    return self::DESCRIPTIONS[$this()];
}
```

2) To allow `$fooBar() === FOO::BAR` in consumer code (perf?).